ui/Model.elm

```diff
@@ -126,42 +154,113 @@ getReleaseList =
         )
         releases
 
 
+processFilterVersions : FilterSet -> List Release -> List Release
+processFilterVersions filterSet releases =
+    List.filter
+        (\{ details } ->
+            List.any
+                (\f -> String.startsWith f details.filename)
+                (Dict.filter (\k v -> v) filterSet |> Dict.keys)
+        )
+        releases
+
+
+filterReleases : Model -> List Release
+filterReleases { filters, releases } =
+    releases
+        |> processFilter filters.channels (.details >> .channel)
+        |> processFilter filters.langs (.details >> .lang)
+        |> processFilter filters.targets (.details >> .target)
+        |> processFilterVersions filters.versions
+
+
 update : Msg -> Model -> ( Model, Cmd Msg )
-update message model =
+update message ({ filters } as model) =
     case message of
         ReleasesFetched result ->
             case result of
                 Ok releases ->
                     { model
                         | releases = releases
-                        , filters = { channels = extractChannels releases }
+                        , filters =
+                            { channels = extractFilterSet (.details >> .channel) releases
+                            , langs = extractFilterSet (.details >> .lang) releases
+                            , targets = extractFilterSet (.details >> .target) releases
+                            , versions = extractVersions releases
+                            }
                     }
                         ! []
 
                 Err err ->
                     Debug.crash "Unhandled Kinto error"
 
         ToggleChannelFilter channel active ->
-            { model
-                | filters =
-                    { channels =
-                        Dict.update channel (\_ -> Just active) model.filters.channels
+            let
+                channels =
+                    Dict.update channel (\_ -> Just active) model.filters.channels
+
+                newFilters =
+                    { filters | channels = channels }
+            in
+                { model | filters = newFilters } ! []
+
+        ToggleLangFilter lang active ->
+            let
+                newFilters =
+                    { filters
+                        | langs = Dict.update lang (\_ -> Just active) filters.langs
+                    }
+            in
+                { model | filters = newFilters } ! []
+
+        ToggleTargetFilter target active ->
+            let
+                newFilters =
+                    { filters
+                        | targets = Dict.update target (\_ -> Just active) filters.targets
+                    }
+            in
+                { model | filters = newFilters } ! []
+
+        ToggleVersionFilter version active ->
+            let
+                newFilters =
+                    { filters
+                        | versions = Dict.update version (\_ -> Just active) filters.versions
                     }
```

Comment:
Yeah, OK. In the Haskell community, there's a thing called "lenses" that people sometimes use for this sort of thing, but they are a bit intimidating for newbies and so they often feel like overkill. Elm doesn't really have any built-in way to do lenses so it's definitely not the right choice here. Maybe we could turn the filters into some kind of two-level structure, where the first level is "version", "lang", etc. and the second level is "50", "51", "52". But maybe that's also more trouble than it's worth. Anyhow I think the change you made here is good 👍